android/modules/database/src/java/ti/modules/titanium/database/DatabaseModule.java

```diff
@@ -53,101 +53,108 @@ public DatabaseModule()
+		if (file instanceof TiFileProxy) {
+			TiFileProxy tiFile = (TiFileProxy) file;
+			String absolutePath = tiFile.getBaseFile().getNativeFile().getAbsolutePath();
+			Log.d(TAG, "Opening database from filesystem: " + absolutePath);
+
+			SQLiteDatabase db = SQLiteDatabase.openDatabase(
+				absolutePath, null, SQLiteDatabase.CREATE_IF_NECESSARY | SQLiteDatabase.NO_LOCALIZED_COLLATORS);
+			if (db != null) {
 				dbp = new TiDatabaseProxy(db);
 			} else {
-				String name = TiConvert.toString(file);
-				SQLiteDatabase db = TiApplication.getInstance().openOrCreateDatabase(name, Context.MODE_PRIVATE, null);
+				throw new RuntimeException("SQLiteDatabase.openDatabase() returned null for path: " + absolutePath);
+			}
+		} else if (file instanceof String) {
+			String name = (String) file;
+			SQLiteDatabase db = TiApplication.getInstance().openOrCreateDatabase(name, Context.MODE_PRIVATE, null);
+			if (db != null) {
 				dbp = new TiDatabaseProxy(name, db);
+			} else {
+				throw new RuntimeException("SQLiteDatabase.openOrCreateDatabase() returned null for name: " + name);
 			}
-
-			Log.d(TAG, "Opened database: " + dbp.getName(), Log.DEBUG_MODE);
-
-		} catch (SQLException e) {
-			String msg = "Error opening database: " + dbp.getName() + " msg=" + e.getMessage();
-			Log.e(TAG, msg, e);
-			throw e;
+		} else if (file != null) {
+			throw new IllegalArgumentException("Ti.Database.open() argument must be of type 'String' or 'File'.");
+		} else {
+			throw new IllegalArgumentException("Ti.Database.open() was given a null argument.");
 		}
 
+		// Return a proxy to the opened database.
+		Log.d(TAG, "Opened database: " + dbp.getName(), Log.DEBUG_MODE);
 		return dbp;
 	}
 
 	@Kroll.method
 	public TiDatabaseProxy install(KrollInvocation invocation, String url, String name) throws IOException
 	{
-		try {
-			Context ctx = TiApplication.getInstance();
-			for (String dbname : ctx.databaseList()) {
-				if (dbname.equals(name)) {
-					return open(name);
-				}
+		// Validate arguments.
+		if ((url == null) || url.isEmpty()) {
+			throw new IllegalArgumentException("Ti.Database.install() 1st argument must be a non-empty string.");
+		}
+		if ((name == null) || name.isEmpty()) {
+			throw new IllegalArgumentException("Ti.Database.install() 2nd argument must be a non-empty string.");
+		}
+
+		// Do not continue if the database has already been installed.
+		// Open a connection to it and stop here.
+		Context ctx = TiApplication.getInstance();
+		for (String dbname : ctx.databaseList()) {
+			if (dbname.equals(name)) {
+				return open(name);
+			}
+		}
+
+		// Fetch a path to the source database file. This is the file to be copied/installed.
+		// Throw an exception if the source database was not found.
+		Log.d(TAG, "db url is = " + url, Log.DEBUG_MODE);
+		if (invocation == null) {
+			throw new RuntimeException("Ti.Database.install() was given a null 'KrollInvocation' object.");
```

Comment:
this seems like "inside baseball" leaking out here. This argument is a special one that the user has no control over and does not deal with and this exception leaking to them may be rather confusing. If this argument is null, then the SDK did something wrong internally.

Reply:
Fair enough. I've modified the code to only resolve/normalize the URL if the `KrollInvocation` object is not `null`.